[Multilib] Extend the Multilib system to support an IncludeDirs field.

simpal01 opened the pull request:
This patch extends the Multilib Yaml format to allow specifying an IncludeDirs for the header path/s per multilib variant. The goal is to enable fine-grained control over header search paths for each multilib variant. This feature is especially useful in setups where header paths deviate from the default bare-metal assumptions. For example, when headers are shared across target triples, it becomes necessary to organize them under target-triple-specific directories to ensure correct resolution.

In the Clang driver, GCCSuffix, OSSuffix and IncludeSuffix are path suffixes that help the compiler locate the correct set of headers and libraries for the selected target. Clang’s multilib infrastructure uses the Multilib class to encapsulate these suffixes. Currently, in the bare-metal Clang driver, all of these suffixes — GCCSuffix, OSSuffix, and IncludeSuffix — are effectively mapped to the Dir field specified in the multilib.yaml configuration.

This patch allows it to be configured independently of Dir, enabling finer-grained control over header search directories for each multilib variant.

Key Changes:

- New Field: IncludeDirs in multilib.yaml (header directory or list of header directories, strings, optional). 

- When present, this path is passed to the Multilib constructor and appended to the sysroot during header path resolution. If omitted, behaviour defaults to preserving current behaviour.

For example,

>  Dir: aarch64-none-elf/aarch64a_exn_rtti_unaligned
>  Flags:
>       --target=aarch64-unknown-none-elf
>  IncludeDirs:
>      include
>      include-libunwind
>      aarch64-none-elf/include
>      aarch64-none-elf/aarch64a_exn_rtti_unaligned/include

Implementation Notes:

1.  Extend the YAML parser to read the IncludeDirs key.
2. Update the Multilib constructor to store the content from this field.
3.  Ensure the driver logic reads and applies IncludeDirs correctly.

simpal01 committed:
[Multilib] Extend the Multilib system to support an IncludepPath field.

This patch extends the Multilib Yaml format to allow
specifying an IncludePath for the header path/s per
multilib variant. The goal is to enable fine-grained
control over header search paths for each multilib variant.
This feature is especially useful in setups where header
paths deviate from the default bare-metal assumptions.
For example, when headers are shared across target triples,
it becomes necessary to organize them under target-triple-specific
directories to ensure correct resolution.

In the Clang driver, GCCSuffix, OSSuffix and IncludeSuffix
are path suffixes that help the compiler locate the correct
set of headers and libraries for the selected target.
Clang’s multilib infrastructure uses the Multilib class to
encapsulate these suffixes. Currently, in the bare-metal
Clang driver, all of these suffixes — GCCSuffix, OSSuffix, and
IncludeSuffix — are effectively mapped to the Dir field specified
in the multilib.yaml configuration.

This patch allows it to be configured independently of Dir,
enabling finer-grained control over header search paths for each multilib variant.

Key Changes:
New Field: IncludePath in multilib.yaml (header path or list of header
paths, strings, optional). When present, this path is passed to the
Multilib constructor and appended to the sysroot during header path resolution.
If omitted, behaviour defaults to preserving current behaviour.

For example,
- Dir: aarch64-none-elf/aarch64a_exn_rtti_unaligned
  Flags:
  - --target=aarch64-unknown-none-elf
  IncludePath:
  - include
  - include-libunwind
  - aarch64-none-elf/include
  - aarch64-none-elf/aarch64a_exn_rtti_unaligned/include

Implementation Notes:
1. Extend the YAML parser to read the IncludePath key.
2. Update the Multilib constructor to store the content from this field.
3. Ensure the driver logic reads and applies IncludePath correctly.

simpal01 committed:
fixup! [Multilib] Add basic tests for new 'IncludePath' field in multilib.yaml.

simpal01 committed:
fixup! fixup! [Multilib] Add basic tests for new 'IncludePath' field in multilib.yaml.

statham-arm commented on clang/test/Driver/baremetal-multilib-includepath-error.yaml:
```diff
@@ -0,0 +1,28 @@
+# (/include) to IncludePath triggers the expected
+# error.
+
+# # RUN: %clang --target=aarch64-none-elf --multi-lib-config=%s %s -### -o /dev/null 2>&1 \
```
I'm a bit surprised that this double `#` doesn't stop the run from happening!

Also slightly surprised that you didn't need a `not` on the command line. Does the clang driver not return a failure status when it finds an error in `multilib.yaml`?

simpal01 replied:
No. It did not give me.
```
-- Testing: 1 tests, 1 workers --
PASS: Clang :: Driver/baremetal-multilib-includepath-error.yaml (1 of 1)
Exit Code: 0

Command Output (stderr):
--
/data_nvme1n1/LLVM-PROJECT/build-driver/bin/clang --target=aarch64-none-elf --multi-lib-config=/data_nvme1n1/LLVM-PROJECT/llvm-project-fork/clang/test/Driver/baremetal-multilib-includepath-error.yaml /data_nvme1n1/LLVM-PROJECT/llvm-project-fork/clang/test/Driver/baremetal-multilib-includepath-error.yaml -### -o /dev/null 2>&1  | /data_nvme1n1/LLVM-PROJECT/build-driver/bin/FileCheck /data_nvme1n1/LLVM-PROJECT/llvm-project-fork/clang/test/Driver/baremetal-multilib-includepath-error.yaml --check-prefix=CHECK-ERROR # RUN: at line 15
+ /data_nvme1n1/LLVM-PROJECT/build-driver/bin/clang --target=aarch64-none-elf --multi-lib-config=/data_nvme1n1/LLVM-PROJECT/llvm-project-fork/clang/test/Driver/baremetal-multilib-includepath-error.yaml /data_nvme1n1/LLVM-PROJECT/llvm-project-fork/clang/test/Driver/baremetal-multilib-includepath-error.yaml -### -o /dev/null
+ /data_nvme1n1/LLVM-PROJECT/build-driver/bin/FileCheck /data_nvme1n1/LLVM-PROJECT/llvm-project-fork/clang/test/Driver/baremetal-multilib-includepath-error.yaml --check-prefix=CHECK-ERROR

--

********************

Testing Time: 0.07s

Total Discovered Tests: 1
  Passed: 1 (100.00%)
```


When running the command manually:
```
$ /data_nvme1n1/LLVM-PROJECT/build-driver/bin/clang --target=aarch64-none-elf --multi-lib-config=/data_nvme1n1/LLVM-PROJECT/llvm-project-fork/clang/test/Driver/baremetal-multilib-includepath-error.yaml /data_nvme1n1/LLVM-PROJECT/llvm-project-fork/clang/test/Driver/baremetal-multilib-includepath-error.yaml -### -o /dev/null 2>&1

/data_nvme1n1/LLVM-PROJECT/llvm-project-fork/clang/test/Driver/baremetal-multilib-includepath-error.yaml:22:3: error: paths must be relative but "/include" starts with "/"
- Dir: aarch64-none-elf/aarch64a_exn_rtti_unaligned
  ^
clang version 21.0.0git (git@github.com:simpal01/llvm-project-fork.git fe8fcbffbfc7c8a16285d3100f089b6528fb1329)
Target: aarch64-unknown-none-elf
Thread model: posix
InstalledDir: /data_nvme1n1/LLVM-PROJECT/build-driver/bin
```

ah! 	This might be because I unintentionally included -###, which only prints the driver’s command sequence without actually performing compilation.

After removing -###, it became necessary to include **not** to ensure the llvm-lit test passed.

Good catch Simon!

simpal01 replied:
i will update the test accordingly

simpal01 committed:
fixup! Remove `-###` and add missing `not` .

The `-###` flag was unintentionally included,
causing the compiler to only print the driver
command sequence without performing compilation.
This masked the need for the `not` command in the llvm-lit test.
After removing `-###`, the missing `not` was added to correctly
expect a failing compile, ensuring the test behaves as intended.

simpal01 committed:
fixup! Using IncludeDirs instead of IncludePath.

simpal01 committed:
fixup! Apply clang format.

simpal01 committed:
fixup! Rename the test files

simpal01 commented:
https://discourse.llvm.org/t/rfc-support-target-triple-specific-include-paths-derived-from-sysroot-via-multilib-yaml-or-driver-logic/86925/6?u=simpal01

simpal01 commented:
@petrhosek  As discussed in this thread  https://github.com/llvm/llvm-project/pull/165321 please have a look on this patch as well

simpal01 committed:
fixup! Addressing the comments received
